iceoryx_hoofs/vocabulary/include/iox/detail/expected_helper.hpp

```diff
@@ -0,0 +1,247 @@
+    explicit err(T&& t) noexcept
+        : value(std::forward<T>(t))
+    {
+    }
+
+    template <typename... Targs>
+    explicit err(Targs&&... args) noexcept
+        : value(std::forward<Targs>(args)...)
+    {
+    }
+
+    T value;
+};
+
+/// @brief helper class to be able to handle 'void' value type specialization
+template <typename ValueType, typename ErrorType>
+class expected_storage
+{
+  public:
+    expected_storage() noexcept = delete;
+
+    template <typename... Targs>
+    explicit expected_storage(in_place_t, Targs&&... args)
+        : data(in_place_index<VALUE_INDEX>(), std::forward<Targs>(args)...)
+    {
+    }
+
+    template <typename... Targs>
+    explicit expected_storage(unexpect_t, Targs&&... args)
+        : data(in_place_index<ERROR_INDEX>(), std::forward<Targs>(args)...)
+    {
+    }
+
+    bool has_value() const
+    {
+        return data.index() == VALUE_INDEX;
+    }
+
+    bool has_error() const
+    {
+        return data.index() == ERROR_INDEX;
+    }
+
+    ValueType& value_unchecked()
+    {
+        return *data.template get_at_index<VALUE_INDEX>();
+    }
+
+    const ValueType& value_unchecked() const
+    {
+        return *data.template get_at_index<VALUE_INDEX>();
+    }
+
+    ErrorType& error_unchecked()
+    {
+        return *data.template get_at_index<ERROR_INDEX>();
+    }
+
+    const ErrorType& error_unchecked() const
+    {
+        return *data.template get_at_index<ERROR_INDEX>();
+    }
+
+  private:
+    static constexpr uint64_t VALUE_INDEX{0};
+    static constexpr uint64_t ERROR_INDEX{1};
+
+    iox::variant<ValueType, ErrorType> data;
+};
+
+/// @brief helper struct to handle 'void' value type specialization
+template <typename ErrorType>
+class expected_storage<void, ErrorType>
+{
+  public:
+    expected_storage() noexcept = delete;
+
+    template <typename... Targs>
+    explicit expected_storage(in_place_t, Targs&&...)
+        : data(in_place_index<VALUE_INDEX>(), DUMMY_VALUE)
+    {
+    }
+
+    template <typename... Targs>
+    explicit expected_storage(unexpect_t, Targs&&... args)
+        : data(in_place_index<ERROR_INDEX>(), std::forward<Targs>(args)...)
+    {
+    }
+
+    bool has_value() const
+    {
+        return data.index() == VALUE_INDEX;
+    }
+
+    bool has_error() const
+    {
+        return data.index() == ERROR_INDEX;
+    }
+
+    void value_unchecked() const
+    {
+        // nothing to do
+    }
+
+    ErrorType& error_unchecked()
+    {
+        return *data.template get_at_index<ERROR_INDEX>();
+    }
+
+    const ErrorType& error_unchecked() const
+    {
+        return *data.template get_at_index<ERROR_INDEX>();
+    }
+
+  private:
+    static constexpr uint64_t VALUE_INDEX{0};
+    static constexpr uint64_t ERROR_INDEX{1};
+
+    using DummyValueType = bool;
+    static constexpr DummyValueType DUMMY_VALUE{true};
+
+    iox::variant<DummyValueType, ErrorType> data;
```

Comment:
I think that should be enough, right?

```suggestion
    variant<DummyValueType, ErrorType> data;
```

Reply:
But it's not needed :)